include/pybind11/pytypes.h

```diff
@@ -298,6 +298,22 @@ class handle : public detail::object_api<handle> {
 
 #ifdef PYBIND11_HANDLE_REF_DEBUG
 private:
+    void throw_gilstate_error(const std::string &function_name) const {
+        fprintf(stderr,
+                "%s is being called while PyGILState_Check() is failing. "
+                "This usually indicates that you are calling pybind11 functions without holding "
+                "the GIL or before "
+                "Python (and/or this Python module) is finished initializing.\n",
```

Comment:
Let's create a wiki page and link to that. That will be much easier to update / maintain.

Reply:
Sounds good.

Could it also be a subsection in the docs? We already have this: https://github.com/pybind/pybind11/blob/7f23e9f3a47cb6b6373467c18962c16dfdd82fe7/docs/advanced/misc.rst#global-interpreter-lock-gil

I'm pasting something I sent as a reply to a Google-internal question about 6 weeks ago (slightly edited), in case you find some of it helpful while you work on this.

If you feel this is too much feature creep, merging this PR with a simple link to #4246 would still seem sufficient to me. We could update the docs or wiki page later.

________

> When exactly is it necessary to acquire GIL when using Python bindings of C++ code

That answer is actually very simple: for **any** call into the [Python C API](https://docs.python.org/3/c-api/index.html).

If you release the GIL before calling into C++, and the C++ never calls back into Python, the situation is still very simple: you reacquire the GIL when the C++ call returns.

It gets more complicated only if the C++ code somehow calls back into Python, because then it first has to acquire the GIL, then release it again. That's still fairly straightforward but ...

... if the C++ code somehow stores Python objects (concretely `PyObject *` in some form or shape) then oversights slip in more easily ...

... especially if it is a `py::object`, because that will call `Py_XINCREF`/`Py_XDECREF` when it is copied/destroyed, maybe as part of a big C++ object copy constructor ...

... and if those objects are in multiple threads, it becomes even more unobvious when that happens ...

... especially if C++ destructors indirectly destroy things like `py::object` in a non-deterministic order at teardown time.

There is one more thing I haven't mentioned yet: when you call into the Python C API, Python may release and reacquire the GIL before returning. I.e. another thread may run through the same section of code you made the call from. You think you have the GIL, but no, not while your C API call is executing, only before and after.

I guess that covers all pitfalls, more or less.

> why simple bindings like `.def("bar", &Foo::Bar)` don't release GIL by default?

That's what PyCLIF does, and it works very well, to my own amazement. But PyCLIF gives you a much tighter controlled environment than pybind11, which let's you do anything you can do with the raw Python C API, and therefore opens up all the pitfalls there can be.

A little bit of good news: PR #4246 provides basic guard rails, not perfect, but it will catch many common oversights.

Reply:
Wow, that's a nicely written email. I tried to incorporate most of it into https://pybind11--4427.org.readthedocs.build/en/4427/advanced/misc.html#common-sources-of-global-interpreter-lock-errors. What do you think?